xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/page/events/PageNotificationEventListener.java

```diff
@@ -0,0 +1,178 @@
+                    && checkXObjectCondition(descriptor, source)
+                    && this.evaluateVelocityTemplate(descriptor.getAuthorReference(),
+                    descriptor.getValidationExpression())) {
+                observationManager.notify(
+                        new PageNotificationEvent(descriptor),
+                        "org.xwiki.platform:xwiki-platform-notifications-api",
+                        source);
+            }
+        }
+    }
+
+    /**
+     * Ensure that the given the source matches what the descriptor needs.
+     * If the source is an instance of XWikiDocument, will check if the document contains the XObject specified in
+     * the descriptor.
+     *
+     * @param descriptor the event descriptor
+     * @param source the event source
+     * @return
+     */
+    private boolean checkXObjectCondition(PageNotificationEventDescriptor descriptor, Object source) {
+        if (StringUtils.isBlank(descriptor.getObjectType())) {
+            // If no XObject is specified in the Object Type field, we don’t need to check the source.
+            return true;
+        } else if (source instanceof XWikiDocument) {
+            XWikiDocument document = (XWikiDocument) source;
+            Map<DocumentReference, List<BaseObject>> documentXObjects = document.getXObjects();
+
+            LocalDocumentReference localXObjectReference =
+                    documentReferenceResolver.resolve(descriptor.getObjectType()).getLocalDocumentReference();
+
+            /*  We can’t create a DocumentReference when only using descriptor.objectType, so we will have to
+                    iterate through the map */
+            for (DocumentReference documentReference : documentXObjects.keySet())  {
+                if (documentReference.getLocalDocumentReference().equals(localXObjectReference)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Evaluate the given velocity template and return a boolean.
+     *
+     * @param userReference a user reference used to build context
+     * @param templateContent the velocity template that should be evaluated
+     * @return true if the template evaluation returned «true» or if the template is empty
+     */
+    private boolean evaluateVelocityTemplate(DocumentReference userReference, String templateContent)
+    {
+        try {
+            // We don’t need to evaluate the template if it’s empty
+            if (StringUtils.isBlank(templateContent)) {
+                return true;
+            }
+
+            Template customTemplate = templateManager.createStringTemplate(templateContent, userReference);
+            XDOM templateXDOM = templateManager.execute(customTemplate);
+
+            WikiPrinter printer = new DefaultWikiPrinter();
+            BlockRenderer renderer =
+                    this.componentManagerProvider.get().getInstance(BlockRenderer.class, "plain");
```

Comment:
OK so now I know why Clement did it this way, because I pointed him to this code to have an inspiration.
So now it would be better to use @Inject instead.